scripts/ont_sync_to_db.py

```diff
@@ -0,0 +1,326 @@
+    if run_name is None:
+        # Define query pattern
+        experiment_name = process.id if not qc else f"QC_{process.id}"
+        sample_name = (
+            sanitize_string(art.name) if not qc else f"QC_{sanitize_string(art.name)}"
+        )
+        position = (
+            art.udf["ONT flow cell position"]
+            if art.udf["ONT flow cell position"] != "None"
+            else "MN19414"
+        )
+
+        pattern = rf"{experiment_name}/{sample_name}/[^/]*_{position}_{art.udf['ONT flow cell ID']}_[^/]*"
+
+        logging.info(
+            f"No run name supplied. Quering the database for run with path pattern '{pattern}'."
+        )
+
+        for row in view.rows:
+            query = row.value["TACA_run_path"]
+            if re.match(pattern, query):
+                matching_rows.append(row)
+
+    # If the run name is supplied, query the database directly
+    else:
+        logging.info(
+            f"Full run name supplied. Quering the database for run '{run_name}'."
+        )
+        for row in view.rows:
+            if run_name == row.key:
+                matching_rows.append(row)
+
+    return matching_rows
+
+
+def write_to_doc(
+    doc: Document, db: Database, process: Process, art: Artifact, args: Namespace
+):
+    """Update a given document with the given artifact's loading information."""
+
+    library_df = get_ont_library_contents(
+        ont_library=art,
+        list_contents=True,
+        print_dataframe=True,
+    )
+
+    # Info to add to the db doc
+    dict_to_add = {
+        "step_name": process.type.name,
+        "step_id": process.id,
+        "timestamp": TIMESTAMP,
+        "operator": process.technician.name,
+        "load_fmol": art.udf["ONT flow cell loading amount (fmol)"],
+        "load_vol": art.udf["Volume to take (uL)"],
+        "library_name": art.name,
+        "library_id": art.id,
+        "sample_data": library_df.to_dict(orient="records"),
+    }
+
+    if "lims" not in doc:
+        doc["lims"] = {}
+    if "loading" not in doc["lims"]:
+        doc["lims"]["loading"] = []
+    doc["lims"]["loading"].append(dict_to_add)
+
+    db[doc.id] = doc
+
+
+def sync_runs_to_db(process: Process, args: Namespace, lims: Lims):
+    """Executive script, called once."""
+
+    # Assert samplesheet, if any, makes sense in the context of the step UDFs
+    assert_samplesheet(process=process, args=args, lims=lims)
+
+    arts: list[Artifact] = [
+        art for art in process.all_outputs() if art.type == "Analyte"
+    ]
+
+    # Assert that only one input is provided for QC runs
+    if "QC" in process.type.name:
+        assert (
+            len(arts) == 1
+        ), "When starting QC sequencing runs, only one input is allowed."
+
+    # Keep track of which artifacts were successfully updated
+    arts_successful = []
+
+    db: Database = get_ONT_db()
+    view: ViewResults = db.view("info/all_stats")
```

Comment:
For convenience in a different project, could you name the db variable to `nanopore_runs_db`? I will suggest this to be a new standard.

Reply:
3c1f13ed4793ef0969f68b73c84533c7d1e87937